API Model Injection

Allcharles opened the pull request:
# API Model Injection

This PR relates to retrieving models from the API before a component loads.

## Changes

- Created `FilterResolver`, `ListResolver`, and `ShowResolver` template classes
- Created resolvers for each api service
- Created error subroute to perform automatic error handling
  - This is similar to the secondary and action subroutes which are not directly affected by the main page
- Updated forms to newer notification system (`toastr`)
- Optimised api requests using new resolver system
  - Previously each component which needed access to a model had to either request it or be passed the model by its parent
  - New system allows requesting of models to flow through the children allowing all subroutes to access the models in use
- Updated menu system to include Category hierarchy
  - This allowed for the categories to define the resolvers for the child pages
  - Child pages can still define page specific resolvers

## Issues
 
- Pages still need to handle models failing to be retrieved
- Disabled the following unit test files (due to PR size):
  - component\about\pages\credits\credits.component.spec.ts
  - component\about\pages\disclaimers\disclaimers.component.spec.ts
  - component\about\pages\ethics\ethics.component.spec.ts
  - component\data-request\data-request.component.spec.ts
  - component\profile\pages\my-edit\my-edit.component.spec.ts
  - component\profile\pages\profile\my-profile.component.spec.ts
  - component\profile\pages\profile\their-profile.component.spec.ts
  - component\profile\pages\their-edit\their-edit.component.spec.ts
  - component\projects\pages\delete\delete.component.spec.ts
  - component\projects\pages\edit\edit.component.spec.ts
  - component\projects\pages\list\list.component.spec.ts
  - component\projects\pages\new\new.component.spec.ts
  - component\projects\pages\permissions\permissions.component.spec.ts
  - component\projects\pages\request\request.component.spec.ts
  - component\security\pages\login\login.component.spec.ts
  - component\send-audio\send-audio.component.spec.ts
  - component\shared\cms\cms.component.spec.ts
  - component\shared\error-handler\error-handler.component.spec.ts
  - component\shared\permissions-shield\permissions-shield.component.spec.ts
  - component\shared\user-badges\user-badge\user-badge.component.spec.ts
  - component\sites\pages\delete\delete.component.spec.ts
  - component\sites\pages\details\details.component.spec.ts
  - component\sites\pages\edit\edit.component.spec.ts
  - component\sites\pages\edit\edit.component.ts
  - component\sites\pages\harvest\harvest.component.spec.ts
  - component\sites\pages\new\new.component.spec.ts

## Closes

Closes #99

Allcharles committed:
Created API_ROOT provider

Allcharles committed:
Extracted API_CONFIG value

Allcharles committed:
Load app config with application

Allcharles committed:
Inserting config data into environment

Allcharles committed:
Removed appConfigService

Allcharles committed:
Removed AppConfigService

Allcharles committed:
Cleanup imports

Allcharles committed:
Initial look at resolvers

Allcharles committed:
Initial look at resolvers

Allcharles committed:
Added resolvers key to pageInfo

Allcharles committed:
Added resolvers key to pageInfo

Allcharles committed:
Inject project model into project list

Allcharles committed:
Inject project model into project list

Allcharles committed:
Created error subroute

Allcharles committed:
Created error subroute

Allcharles committed:
Created site resolvers

Allcharles committed:
Created site resolvers

Allcharles committed:
Added api service resolvers

Allcharles committed:
Added api service resolvers

Allcharles committed:
Adjusting some pages to new system

Allcharles committed:
Adjusting some pages to new system

Allcharles committed:
Made forms full-sized by default

Allcharles committed:
Added form size option

Re-organised form unit tests

Allcharles committed:
Minor form unit test cleanup

Allcharles committed:
Extracted CMS pages to environment

Allcharles committed:
Fixed unit tests

Allcharles committed:
Linted project

Allcharles committed:
Working form guard for new project page

Allcharles committed:
Cleanup comments

Allcharles committed:
Added WithFormCheck to form pages

atruskie committed:
Merge pull request #114 from QutEcoacoustics/api-service-injectors

Api service injectors

Allcharles committed:
Fix unit tests

atruskie commented on src/app/services/baw-api/account.service.ts:
```diff
@@ -36,3 +37,25 @@ export class AccountService extends ReadonlyApi<User, []> {
+      params => parseInt(params.get("accountId"), 10),
+      () => []
+    );
+  }
```
Okay, that's a lot of ceremony still.

I'd like to see this reduced to a function. On usage something like `ListResolver<User>(AccountsService)`

atruskie reviewed (changes requested):
I'd like to see the resolver functions pretty much user defaults rather than defining a new XXXResolver class for every case. They have almost the same functionality - the core goal is to cut down on boilerplate.

Allcharles committed:
Merge pull request #117 from QutEcoacoustics/full-sized-forms

Full sized forms

Allcharles committed:
Added unit tests

Allcharles committed:
Update data-request.component.spec.ts

Allcharles committed:
Merge pull request #122 from QutEcoacoustics/fix-merge-conflict

Fix Merge Issue

Allcharles committed:
Merge branch 'master' into extract-cms-links

Allcharles committed:
Updated CMS to utilise new environment

Allcharles committed:
Passing unit tests

Allcharles committed:
Removed CMS_DATA token

Allcharles committed:
Address PR comments

Allcharles committed:
Merge branch 'master' into form-guard

Allcharles committed:
Merge pull request #118 from QutEcoacoustics/extract-cms-links

Extract cms links

Allcharles committed:
Merge branch 'master' into form-guard

Allcharles committed:
Revert "Merge branch 'master' into form-guard"

This reverts commit 607a9f7bac3285ef52af8ca057c56014ca812296.

Allcharles committed:
Merge branch 'master' into form-guard

Allcharles committed:
Manual MERGE from master

Allcharles committed:
Fixed unit tests

Allcharles committed:
Merge pull request #121 from QutEcoacoustics/form-guard

Form guard

Allcharles committed:
Minor optimisations to boilerplate

Allcharles committed:
Minor optimisations to boilerplate

Allcharles committed:
Reducing size of combined resolvers

Allcharles committed:
Reducing size of combined resolvers

Allcharles committed:
Reduced size of inidividual resolvers

Allcharles committed:
Reduced size of inidividual resolvers

Allcharles committed:
Removed angular material

Allcharles committed:
Reduced unsubscribe boilerplate

Allcharles committed:
Abstract unsubscribe code to mixin

Allcharles committed:
Linted project

Allcharles committed:
Patch broken unit tests

Allcharles committed:
Defined WithUnsubscribe test headers

Allcharles committed:
Implemented unit tests

Allcharles committed:
Linted project

Allcharles committed:
Cleanup floating OnDestroy functions

Allcharles committed:
Created DeploymentEnvironmentService

- Created DeploymentEnvironmentService
- Added multiple fetch deployment environment attempts
- Added error message when deployment environment fails to load

Allcharles committed:
Cleanup typing

Allcharles committed:
Changed DeploymentEnvironment Service to appConfigService

Allcharles committed:
Fixed unit tests

Allcharles committed:
Fixed unit tests

Removed global services, this forces unit tests to actually utilise the helper functions which means mock services run instead of the dangerous live versions.

Allcharles committed:
Linted project

atruskie commented on src/app/services/baw-api/resolver-common.ts:
```diff
@@ -1,53 +1,124 @@
+
+    return [
+      {
+        provide: name + "sResolver",
```
I wonder if you could use `typeof` here to get typescript to put the name is for us?

Also, don't bother with the pluralization... it's not worth it in the end. Why not simpler: `XXXListResolver`

Allcharles replied:
Unfortunately `typeof` does not work on custom classes, only `instanceof`

atruskie reviewed (approved):
Much cleaner implementation.

I'd suggest adding a few comments around the resolvers-common file to remind us of choices made and general workflow.

atruskie committed:
Merge pull request #124 from QutEcoacoustics/extract-angular-material

Removed angular material dependency

Allcharles committed:
Merge pull request #125 from QutEcoacoustics/unsubscribe-mixin

Unsubscribe mixin

Allcharles committed:
Merge branch 'master' into app-deploy-environment

Allcharles committed:
Merge branch 'master' into app-deploy-environment

Allcharles committed:
Merge pull request #126 from QutEcoacoustics/app-deploy-environment

App deploy environment

Allcharles committed:
Changed resolver names

Allcharles committed:
Changed resolver names

Allcharles committed:
Updated documentation

Allcharles committed:
Updated documentation

Allcharles committed:
Linted project

Allcharles committed:
Merge branch 'api-model-injection' of https://github.com/Allcharles/AngularTestbench into api-model-injection

Allcharles committed:
Updated profile pages

Allcharles committed:
Reduced duplication of resolvers list

Allcharles committed:
Created category parent tree

Allcharles committed:
Updated site pages

Allcharles committed:
Updated project pages

Allcharles committed:
Updated form submissions

Allcharles committed:
Cleanup project details unit tests

Allcharles committed:
Disabled failing unit tests

Allcharles committed:
Lint project
